cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/MetadataClient.java

```diff
@@ -4,123 +4,164 @@
-    if (_time.milliseconds() >= _metadata.lastSuccessfulUpdate() + _metadataTTL) {
-      // Cruise Control always fetch metadata for all the topics.
-      _metadata.needMetadataForAllTopics(true);
-      int version = _metadata.requestUpdate();
-      long remaining = timeout;
-      Cluster beforeUpdate = _metadata.fetch();
-      boolean isMetadataUpdated = _metadata.version() > version;
-      while (!isMetadataUpdated && remaining > 0) {
-        _metadata.requestUpdate();
-        long start = _time.milliseconds();
-        _networkClient.poll(remaining, start);
-        remaining -= (_time.milliseconds() - start);
-        isMetadataUpdated = _metadata.version() > version;
+    if (_time.milliseconds() >= _lastSuccessfulUpdate + _metadataTTL) {
+      // Cruise Control always fetches metadata for all the topics.
+      boolean isMetadataUpdated = false;
+      Optional<Cluster> refreshedCluster = Optional.empty();
+      try {
+        refreshedCluster = Optional.of(doRefreshMetadata(timeout));
+        isMetadataUpdated = true;
+      } catch (ExecutionException | TimeoutException | InterruptedException e) {
+        LOG.warn("Exception while updating metadata", e);
       }
-      if (isMetadataUpdated) {
+
+      if (isMetadataUpdated && refreshedCluster.isPresent()) {
+        _lastSuccessfulUpdate = _time.milliseconds();
+        _version += 1;
         if (LOG.isDebugEnabled()) {
-          LOG.debug("Updated metadata {}", _metadata.fetch());
+          LOG.debug("Updated metadata {}", _cluster);
         }
-        if (MonitorUtils.metadataChanged(beforeUpdate, _metadata.fetch())) {
+        Cluster newCluster = refreshedCluster.get();
+        if (MonitorUtils.metadataChanged(_cluster, newCluster)) {
           _metadataGeneration.incrementAndGet();
+          _cluster = newCluster;
         }
       } else {
         LOG.warn("Failed to update metadata in {}ms. Using old metadata with version {} and last successful update {}.",
-                 timeout, _metadata.version(), _metadata.lastSuccessfulUpdate());
+                 timeout, _version, _lastSuccessfulUpdate);
       }
     }
-    return new ClusterAndGeneration(_metadata.fetch(), _metadataGeneration.get());
+    return new ClusterAndGeneration(_cluster, _metadataGeneration.get());
   }
 
-  /**
-   * Close the metadata client. Synchronized to avoid interrupting the network client during a poll.
-   */
-  public synchronized void close() {
-    _networkClient.close();
+  private Cluster doRefreshMetadata(int timeout) throws InterruptedException, ExecutionException, TimeoutException {
+    // We use both AdminClient timeout options and `future.get(timeout)`
+    // The former ensures the timeouts are enforced in the brokers
+    int remaining = timeout;
+    long start = _time.milliseconds();
+
+    Set<String> topicNames = _adminClient.listTopics(new ListTopicsOptions().timeoutMs(remaining))
+        .names().get(remaining, TimeUnit.MILLISECONDS);
```

Comment:
Overall timeout can potentially be more than `remaining`, because 
1) `_adminClient.listTopics(new ListTopicsOptions().timeoutMs(remaining))` can take up to `remaining` time, and
2) `.names().get(remaining, TimeUnit.MILLISECONDS);` can take up to remaining timeout, as well.

Reply:
I don't think that's the case because `_adminClient.listTopics(new ListTopicsOptions().timeoutMs(remaining))` simply enqueues the call in the background. The timeout there tells the background thread to give up after `remaining` (+ also pass the timeoutMs in the request, if applicable)

`.names().get(remaining, TimeUnit.MILLISECONDS);` is the blocking code which will not take more than `remaining`